independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfiguratorBase.java

```diff
@@ -134,6 +142,74 @@ public THIS addType(Class<?> type) {
         return addType(DotName.createSimple(type.getName()));
     }
 
+    /**
+     * Adds an unrestricted set of bean types for the given type as if it represented a bean class of a managed bean.
+     *
+     * @param typeName {@link DotName} representation of a class that should be scanned for types
+     * @return self
+     */
+    public THIS addTypeClosure(DotName typeName) {
+        this.registeredTypeClosures.add(typeName);
+        return self();
+    }
+
+    /**
+     * Adds an unrestricted set of bean types for the given type as if it represented a bean class of a managed bean.
+     *
+     * @param type a class that should be scanned for types
+     * @return self
+     */
+    public THIS addTypeClosure(Class<?> type) {
+        return addTypeClosure(DotName.createSimple(type));
+    }
+
+    /**
+     * Adds an unrestricted set of bean types for the given type as if it represented a bean class of a managed bean.
+     *
+     * @param type {@link Type} representation of a class that should be scanned for types
+     * @return self
+     */
+    public THIS addTypeClosure(Type type) {
+        return addTypeClosure(type.name());
+    }
+
+    /**
+     * Removes listed types from the resulting types of the synthetic bean.
+     *
+     * @param types types that should be removed from the resulting set of bean types
+     * @return self
+     */
+    public THIS restrictTypes(Class<?>... types) {
+        for (Class<?> classType : types) {
+            restrictTypes(Type.create(classType));
+        }
+        return self();
+    }
+
+    /**
+     * Removes listed types from the resulting types of the synthetic bean.
+     *
+     * @param types types that should be removed from the resulting set of bean types
+     * @return self
+     */
+    public THIS restrictTypes(DotName... types) {
```

Comment:
Note that this is different from the discussion in https://github.com/quarkusio/quarkus/issues/44887.

Instead of listing all the types you want to end up with, I choose to instead list the types that you want removed from the set of types. My reasoning is that this is mostly useful when combined with `addTypeClosure` where you want to add bigger amount of types and here you just want to exclude a few.
If you need to list an excessive list of types here, you might as well not register the whole closure and have these types under `addTypes` directly.

Reply:
In that case, it might make more sense to rename the method to something like `removeTypes()`.

Reply:
Agree with `removeTypes()`. [Restricting the bean types of a bean](https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1.html#restricting_bean_types) is a well defined operation and this is not it.

Reply:
Also, if this was `restrictTypes()` as defined by the spec, I would disagree with adding the overload accepting `Type...`. Restriction accepts a set of classes and keeps only the types whose erasures are included in the set. This is important in case of parameterized types.

With `removeTypes()`, we can be more direct and expect users to explicitly remove parameterized types if they want.

Reply:
> In that case, it might make more sense to rename the method to something like `removeTypes()`.

Agreed, that's definitely a more fitting name.